lib/openssl/buffering.rb

```diff
@@ -345,15 +345,17 @@ def do_write(s)
     @wbuffer << s
     @wbuffer.force_encoding(Encoding::BINARY)
     @sync ||= false
-    if @sync or @wbuffer.size > BLOCK_SIZE
-      until @wbuffer.empty?
+    buffer_size = @wbuffer.size
+    if @sync or buffer_size > BLOCK_SIZE
+      nwrote = 0
+      while nwrote < buffer_size do
         begin
-          nwrote = syswrite(@wbuffer)
+          nwrote += syswrite(@wbuffer)
         rescue Errno::EAGAIN
           retry
         end
-        @wbuffer[0, nwrote] = ""
       end
+      @wbuffer[0, nwrote] = ""
```

Comment:
I'm a little concerned that if `syswrite` raises some other error, `@wbuffer` will not be truncated correctly. WDYT?

Reply:
I think you're right here - it needs a general exception handler to truncate the buffer and raise in that case.  I'll add that later today.

Reply:
The buffer clear is now within an ensure, so any written bytes will always be cleared - thanks for catching this!